contrib/babelfishpg_tsql/src/pltsql_utils.c

```diff
@@ -1983,7 +1985,173 @@ check_fulltext_exist(const char *schema_name, const char *table_name)
+				(errcode(ERRCODE_INVALID_COLUMN_REFERENCE),
+					errmsg("Cannot use a CONTAINS or FREETEXT predicate on column \'%s\' because it is not full-text indexed.",
+						(char *)(column_name->elements[i]).ptr_value)));
+		}
+	}
+	list_free(ft_column_name);
+	return true;
+}
+
+ 
+/* Check if the columns provided in the freetext predicate query exist or not
+ * The check for the columns existence is handled in BISON, but we require to 
+ * check it before checking if they are fulltext indexed or not
+ */
+void check_column_list(Oid relid, const List *column_name)
+{
+ 	Relation 	relation = RelationIdGetRelation(relid);
+ 	TupleDesc	tupledesc = RelationGetDescr(relation);
+ 
+ 	for(int i = 0; i < column_name->length; i++)
+ 	{
+ 		bool flag = false;
+ 		for (int j = 0; j < tupledesc->natts; j++) 
+ 		{
+ 			Form_pg_attribute attr = TupleDescAttr(tupledesc, j);
+ 			if (!attr->attisdropped) 
+ 			{ 
+ 				char *str = NameStr(attr->attname);
+ 				if(strcmp(toLower((char *)(column_name->elements[i]).ptr_value), toLower(str)) == 0)
+ 				{
+ 					flag = true;
+ 					break;
+ 				}
+ 			}
+ 		}
+ 		if(!flag)
+ 		ereport(ERROR,
+ 			(errcode(ERRCODE_INVALID_OBJECT_DEFINITION),
+ 				errmsg("column \"%s\" does not exist.",
+ 					(char *)(column_name->elements[i]).ptr_value)));
+ 	}
+	
+	RelationClose(relation);
+	return;
+}
+ 
+List
+*get_fulltext_indexed_columns(Oid relid, char *ft_index_name)
+{
+ 	Relation        relation  = RelationIdGetRelation(relid);
+ 	List            *indexoidlist = RelationGetIndexList(relation);
+ 	List 		*column_name = NULL;
+ 	ListCell	*cell;
+ 	char		*idx;
+ 	Datum 		result;
+ 
+ 	foreach(cell, indexoidlist)
+ 	{
+ 		Oid indexOid = lfirst_oid(cell);
+ 		char *name = get_rel_name(indexOid);
+ 
+ 		if (strcmp(name, ft_index_name) == 0)
+ 		{
+ 			result = DirectFunctionCall1(pg_get_indexdef, ObjectIdGetDatum(indexOid));
+ 			if(DatumGetPointer(result) == NULL)
+ 				return NULL;
+ 			idx = text_to_cstring(DatumGetTextP(result));
+ 			column_name = get_columns(idx);
+ 			break;
+ 		}
+ 	}
+
+	RelationClose(relation);
+	list_free(indexoidlist);
+
+     	return column_name;
+}
+ 
+/* The columns over which GIN index is created the index definition is like:
+ * CREATE INDEX ft_table_name ON table_name USING gin (to_tsvector('fts_contains_simple'::reconfig, column_name)), for babelfish version before 16.2
+ * and
+ * CREATE INDEX ft_table_name ON table_name USING gin (to_tsvector('fts_contains_simple'::reconfig, replace_special_chars_fts(column_name))), for babelfish version 16.2 onwards
+ * These columns can be of any text based datatype, like: TEXT, CHAR, VARCHAR, VARBINARY, etc.
+ */
+List
+*get_columns(char *index_stmt)
+{
+ 	List	 *column_name = NULL;
+ 	// Find the starting point after "to_tsvector"
+ 	const char* indexdf = index_stmt;
+ 
+ 	while ((indexdf = strstr(indexdf, "regconfig, ")) != NULL) 
+ 	{
+ 		int i = 0;
+ 		StringInfoData bufStr;
+ 
+ 		initStringInfo(&bufStr);
+ 		indexdf += strlen("regconfig, ");
```

Comment:
We can calculate length of `"regconfig, "` and `"replace_special_chars_fts("` once and reuse them.